src/mem/threadalloc.h

```diff
@@ -167,10 +182,15 @@ namespace snmalloc
 #  ifdef USE_MALLOC
       return get_reference();
 #  else
-      auto alloc = get_reference();
+      auto*& alloc = get_reference();
       if (unlikely(needs_initialisation(alloc)))
       {
-        alloc = reinterpret_cast<Alloc*>(init_thread_allocator());
+        std::function<void*(void*)> f = [](void*) { return nullptr; };
+        // Call `init_thread_allocator` to perform down call in case
+        // register_clean_up does more.
+        // During teardown for the destructor based ThreadAlloc this will set
+        // alloc to GlobalPlaceHolder;
+        init_thread_allocator(f);
```

Comment:
"The returned Alloc* is guaranteed to be initialised" is not true during teardown right? I reckon we can drop that guarantee, I doubt this is a useful guarantee anyway.